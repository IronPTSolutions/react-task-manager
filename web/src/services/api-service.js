import axios from "axios";

const service = axios.create({
  withCredentials: true,
  baseURL: import.meta.env.VITE_BASE_API_URL || "http://localhost:3000/v1",
});

service.interceptors.response.use(
  (response) => {
    return response.data;
  },
  (error) => {
    if (
      error.response.status === 401 &&
      window.location.pathname !== "/login"
    ) {
      localStorage.removeItem("user");
      window.location.assign("/login");
    } else {
      return Promise.reject(error);
    }
  }
);

export function login(data) {
  return service.post("/login", data);
}

export function getGroups() {
  return service.get("/task-groups");
}

export function getTasks() {
  return service.get("/tasks");
}

export function getTask(id) {
  return service.get(`/tasks/${id}`);
}

export function deleteTask(id) {
  return service.delete(`/tasks/${id}`);
}

export function createTask(body) {
  return service.post("/tasks", body);
}

export function editTask(id, body) {
  return service.patch(`/tasks/${id}`, body);
}

export function logoutApi() {
  return service.post("/logout");
}

export function createUser(body) {
  const formData = new FormData();

  formData.append("name", body.name);
  formData.append("email", body.email);
  formData.append("password", body.password);

  if (body.avatar) {
    formData.append("avatar", body.avatar[0]);
  }

  return service.post("/users", formData);
}
