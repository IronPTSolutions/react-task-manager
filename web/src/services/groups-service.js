import axios from "axios";

const service = axios.create({
  withCredentials: true,
  baseURL: import.meta.env.REACT_APP_BASE_API_URL || "http://localhost:3000/v1",
});

export function login(data) {
  return service.post("/login", data).then((response) => response.data);
}

export function getTasks() {
  return service.get("/tasks").then((response) => response.data);
}
