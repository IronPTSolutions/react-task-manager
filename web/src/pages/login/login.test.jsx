import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { beforeEach, describe, expect, test, vi } from "vitest";
import LoginPage from "./login";
import * as authContext from "../../contexts/auth-context";
import * as apiService from "../../services/api-service";
import { BrowserRouter } from "react-router-dom";

describe("<Login/>", () => {
  beforeEach(() => {
    cleanup();
  });

  test("happy case: render OK", () => {
    // Given
    vi.spyOn(authContext, "useAuthContext").mockReturnValue({
      onLogin: vi.fn(),
    });

    // When
    render(<LoginPage />, { wrapper: BrowserRouter });

    // Then
    expect(screen.getAllByText("Login")).toBeTruthy();
    expect(screen.getByLabelText("Email address")).toBeTruthy();

    expect(document.body).toMatchSnapshot();
  });

  test("happy case", async () => {
    // Given
    const onLogin = vi.fn();

    vi.spyOn(authContext, "useAuthContext").mockReturnValue({
      onLogin,
    });

    const loginServiceMock = vi.spyOn(apiService, "login").mockResolvedValue({
      response: {},
    });

    // When
    render(<LoginPage />, { wrapper: BrowserRouter });

    const email = screen.getByLabelText("Email address");
    const password = screen.getByLabelText("Password");

    fireEvent.change(email, { target: { value: "john@example.com" } });
    fireEvent.change(password, { target: { value: "123" } });

    const form = screen.getByTestId("login-form");

    fireEvent.submit(form);

    // Then
    await waitFor(() => {
      expect(loginServiceMock).toHaveBeenCalledWith({
        email: "john@example.com",
        password: "123",
      });

      expect(onLogin).toBeCalledWith({ response: {} });
    });
  });
});
