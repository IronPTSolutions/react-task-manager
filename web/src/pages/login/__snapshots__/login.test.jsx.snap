// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Login/> > happy case: render OK 1`] = `
<body>
  <div>
    <div>
      <h1>
        Login
      </h1>
      <form
        data-testid="login-form"
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="email"
          >
            Email address
          </label>
          <input
            class="form-control"
            id="email"
            name="email"
            type="email"
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="exampleInputPassword1"
          >
            Password
          </label>
          <input
            class="form-control"
            id="exampleInputPassword1"
            name="password"
            type="password"
          />
        </div>
        <button
          class="btn btn-primary"
          type="submit"
        >
          Login
        </button>
        <a
          class="btn btn-link"
          href="/signup"
        >
          Sign up
        </a>
      </form>
    </div>
  </div>
</body>
`;
